internals/overlord/planstate/manager.go

```diff
@@ -0,0 +1,192 @@
+package planstate
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+// LabelExists is the error returned by AppendLayer when a layer with that
+// label already exists.
+type LabelExists struct {
+	Label string
+}
+
+func (e *LabelExists) Error() string {
+	return fmt.Sprintf("layer %q already exists", e.Label)
+}
+
+type PlanManager struct {
+	state     *state.State
+	runner    *state.TaskRunner
+	pebbleDir string
+
+	planLock     sync.Mutex
+	plan         *plan.Plan
+	planHandlers []PlanChangeFunc
+}
+
+func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string) (*PlanManager, error) {
+	manager := &PlanManager{
+		state:     s,
+		runner:    runner,
+		pebbleDir: pebbleDir,
+	}
+
+	return manager, nil
+}
+
+// Load reads plan layers from the pebble directory, combine and validate the
+// final plan, and finally notifies registered managers up the plan update.
+func (m *PlanManager) Load() error {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+	plan, err := plan.ReadDir(m.pebbleDir)
+	if err != nil {
+		return err
+	}
+	m.planChanged(plan)
+	return nil
+}
+
+// PlanChangeFunc is the type of function used by AddChangeListeners.
+type PlanChangeFunc func(p *plan.Plan)
+
+// AddChangeListeners adds f to the list of functions that are called whenever
+// the plan has changed. This method may not be called once the overlord state
+// engine has started.
+func (m *PlanManager) AddChangeListeners(f PlanChangeFunc) {
+	m.planHandlers = append(m.planHandlers, f)
+}
+
+func (m *PlanManager) planChanged(plan *plan.Plan) {
+	m.plan = plan
+	for _, f := range m.planHandlers {
+		f(plan)
+	}
+}
+
+// Plan returns the configuration plan. Any change made to the plan will
+// result in a new Plan instance, so the current design assumes a returned
+// plan is never mutated by planstate.
+func (m *PlanManager) Plan() (*plan.Plan, error) {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+	return m.plan, nil
+}
+
+// LayerCreator allows an Append or Combine operation on the plan to include
+// layer creation derived from the plan itself, which has to happen while
+// the plan is locked.
+type LayerCreator interface {
+	Layer(planIn *plan.Plan) (*plan.Layer, error)
+}
```

Comment:
The only real functional addition to the plan related code moved out of the Service Manager into the Plan Manager.

Reply:
Removed.